### DISCUSSION TITLE
Library to track Druid Eclipse - just needs to be integrated

### DISCUSSION THREAD

**Discussion by Volgrath** [Sidoine/Ovale#789] [Show and tell]
```lua
local __exports = LibStub:NewLibrary("ovale/DruidEclipse", 80300)
if not __exports then return end
local __class = LibStub:GetLibrary("tslib").newClass
local aceEvent = LibStub:GetLibrary("AceEvent-3.0", true)
local GetTime = GetTime
local CombatLogGetCurrentEventInfo = CombatLogGetCurrentEventInfo
local ECLIPSE_SOLAR_BUFF_ID = 48517
local ECLIPSE_SOLAR_RESET_SPELLS = {
	[194223] = true,
	[102560] = true,
	[48518] = true
}
local ECLIPSE_LUNAR_BUFF_ID = 48518
local ECLIPSE_LUNAR_RESET_SPELLS = {
	[194223] = true,
	[102560] = true,
	[48517] = true
}
__exports.OvaleDruidEclipseClass = __class(nil, {
    constructor = function(self, ovaleAura, ovale)
	self.ovaleAura = ovaleAura
	self.ovale = ovale
	self.eclipseLunarOff = false
	self.eclipseSolarOff = false
        self.OnInitialize = function()
            if self.ovale.playerClass == "DRUID" then
                self.module:RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED", self.COMBAT_LOG_EVENT_UNFILTERED)
            end
        end
        self.OnDisable = function()
            if self.ovale.playerClass == "DRUID" then
                self.module:UnregisterEvent("COMBAT_LOG_EVENT_UNFILTERED")
            end
        end
        self.COMBAT_LOG_EVENT_UNFILTERED = function(event, ...)
            local _, subtype, _, sourceGUID, _, _, _, _, _, _, _, spellID = CombatLogGetCurrentEventInfo()
            local me = self.ovale.playerGUID
            if sourceGUID == me then
                if subtype == "SPELL_CAST_SUCCESS" and ECLIPSE_LUNAR_RESET_SPELLS[spellID] then
                    self:SetLunarEclipseReady(GetTime())
                end
                if subtype == "SPELL_CAST_SUCCESS" and ECLIPSE_SOLAR_RESET_SPELLS[spellID] then
                    self:SetSolarEclipseReady(GetTime())
                end
            end
        end
        self.module = ovale:createModule("OvaleDruidEclipse", self.OnInitialize, self.OnDisable, aceEvent)
    end,
    SetLunarEclipseReady = function(self, atTime)
        self.eclipseLunarOff = false
    end,
    SetSolarEclipseReady = function(self, atTime)
        self.eclipseSolarOff = false
    end,
    EclipseLunarOff = function(self, atTime)
        local active = self:GetEclipseLunarActive(atTime)
        if active then
            self.eclipseLunarOff = true
        end
        return self.eclipseLunarOff
    end,
    EclipseSolarOff = function(self, atTime)
        local active = self:GetEclipseSolarActive(atTime)
        if active then
            self.eclipseSolarOff = true
        end
        return self.eclipseSolarOff
    end,
    GetEclipseLunarActive = function(self, atTime)
        local aura = self.ovaleAura:GetAura("player", ECLIPSE_LUNAR_BUFF_ID, atTime, "HELPFUL", true)
        local active = aura and self.ovaleAura:IsActiveAura(aura, atTime) or 0
        return active
    end,
    GetEclipseSolarActive = function(self, atTime)
        local aura = self.ovaleAura:GetAura("player", ECLIPSE_SOLAR_BUFF_ID, atTime, "HELPFUL", true)
        local active = aura and self.ovaleAura:IsActiveAura(aura, atTime) or 0
        return active
    end,
})
```

**Comment by Volgrath** [comment]
For when testing integration - This is intended to have EclipseLunarOff() and EclipseSolarOff() be functions you can call within the Ovale code.  These return true when you CANNOT cast Wrath/Starfire toward building the appropriate Eclipse (out of "charges"), and false when you can (they have 1+ "charges").

**Reply by johnnylam88** [reply]
Unfortunately, the way that you chose to track Eclipse state isn't how it works in game. There's more to it than just checking whether the buffs are active and going in the other direction once you've entered either Eclipse state. Other things that need to be handled are:

- Celestial Alignment causes both Eclipses to become active.
- At the start of combat and when exiting CA, you may enter either Eclipse.
- After leaving combat for more than ~5 seconds, you may enter either Eclipse.
- In Torghast, you may gain an anima power where you may enter Eclipse after a single cast.

I'm actively working on a druid module that models this correctly and I will submit a pull-up for it hopefully this week.

**Reply by johnnylam88** [reply]
Just to correct my reading of the code now that I'm at my computer instead of on my phone, I do see that you've incorporated CA (and Incarnation) for Eclipse state resets.

**Comment by Volgrath** [comment]
Just to clarify, the code is more to be able to suggest when wrath/starfire are available to build toward an eclipse rather than the current Eclipse state as that buff can already be tracked.  The druid rotation code would then need to track any Eclipse buff(s) and the result of this code separately to recommend wrath/starfire where appropriate.
    On Tuesday, July 13, 2021, 01:48:36 PM MDT, Johnny C. Lam ***@***.***> wrote:  
 
 


Just to correct my reading of the code now that I'm at my computer instead of on my phone, I do see that you've incorporated CA (and Incarnation) for Eclipse state resets.

**Reply by johnnylam88** [reply]
What you're trying to do here can be much more directly done by querying `GetSpellCount(WRATH_SPELLID)` and `GetSpellCount(STARFIRE_SPELLID)` and using their values to determine the correct state without needing to catch CLEU events.

**Reply by Volgrath** [reply]
I got mixed results trying to incorporate that into the druid rotation back then.  Maybe the GetSpellCount() has improved since then.

**Comment by johnnylam88** [comment]
This is a addressed by https://github.com/Sidoine/Ovale/pull/952 that adds Eclipse state tracking to Ovale, and provides the SimC-equivalent script conditions: `EclipseAnyNext`, `EclipseLunarNext`, `EclipseSolarNext`, `EclipseLunarIn()`, and `EclipseSolarIn()`.
